### DISCUSSION TITLE
Feeding MediaPlayer from an external queue or BytesIO buffer

### DISCUSSION THREAD

**Discussion by roman-verbit-ai** [aiortc/aiortc#656] [Q&A]
Hi, 
Thanks for this great project - it has been a key enabler for us!

I was wondering if you could advice regarding the recommended way to work with an external source of media.
In our scenario, the media is incoming from another component, chunks of raw media are being pushed to a Queue.
What we want to do is take these chunks and transmit them over WebRTC to a remote peer (browser) who will consume this media.

So, we cannot simply do `av.open(???)` to tap in the real source of the media.
After some playing around, I managed to use a python `BytesIO` buffer as the source, like this: `av.open(bytes_io_buffer)`
We need to write the contents on the incoming queue into the buffer, and then let `MediaPlayer` read it and do its magic.
The media is incoming in realtime, we do not have the entire stream/file available in advance.

There are various issues here concerning buffer underrun:
1. Starting with an empty buffer, the call to `av.open(bytes_io_buffer)` fails - it expects at least something to be inside it.
2. We might reach some points where the buffer will have no new data - and then I expect trying to decode the next frame will misbehave:
```
while not quit_event.is_set():
        try:
            frame = next(container.decode(*streams))
```
https://github.com/aiortc/aiortc/blob/7ecc99ab77b3c1e70f1e67770517d6334cf51f57/src/aiortc/contrib/media.py#L111

My question is: 
What is the recommended method to feed MediaPlayer from an external in-memory source of media?
Any method will do, we have no problem with subclassing MediaPlayer - but not sure which approach is best in terms of the internal works of PyAV and all other functionality implemented in MediaPlayer currently.

**Comment by 0x5F0887DF** [comment]
Hi @roman-verbit-ai 
Have you found the solution?

I'm not sure that it's a true way, but I've used the MediaRelay.

https://github.com/aiortc/aiortc/blob/02ff89e4040820c2ea7dae483452f20e83db4557/src/aiortc/contrib/media.py#L451-L457

The code looks like
```
def start_relay():
    relay = MediaRelay()
    track = YourStreamTrack()
    proxy: RelayStreamTrack = relay.subscribe(track)
    relay._start(proxy)
```

where `YourStreamTrack` is subclass of `MediaStreamTrack`

**Reply by roman-verbit-ai** [reply]
@i-spirin 
As I mentioned, we ended up replacing `MediaPlayer` with our own version where `av.open()` does not happen so early.
Instead, we push the incoming media into a queue, and instantiate the AV container only when there's "enough" data in the queue.

This is far from perfect, because we basically need to guess what is "enough" for every possible container format.
Also, some formats don't work at all because they require the entire file to be loaded/seekable (e.g. mp4).

I did not really understand how the `MediaRelay` that you mentioned is relevant and how is should be used?
If I need to subclass my own `YourStreamTrack` - I need to generate `Frame` objects on my own somehow...
So, all of my problems are earlier than this - I need to demux incrementally from a stream - without having the whole file in my hands.

I realize this is much more a PyAV issue than an `aiortc` one. 
However, it seems like the MediaPlayer and its satellite objects should have given a solution to it - just feels like I didn't find it.
I don't believe streaming media from a custom application source is such a rare use case... is it?
(as opposed to letting av do it via `av.open("magic url")`)

**Comment by anthnyprschka** [comment]
@roman-verbit-ai I tried your approach, and read full length MP4 into BytesIO, which I then pass into MediaPlayer. However, I get undefined or unsupported codec error. Any ideas?

**Comment by ajiansoft** [comment]
Hey, guys, is there a solution to this problem now?
